File: ruby/examples/simple_robot.rb
#!/usr/bin/env ruby

require_relative '../lib/eywa'

def process_task
  # Get current task
  task_thread = get_task
  task = task_thread.value
  
  info("Processing task", {
    task_id: task["euuid"],
    message: task["message"]
  })
  
  # Update status
  update_task(PROCESSING)
  
  # Simulate some work
  info("Starting data processing...")
  sleep(1)
  
  # Report progress
  report("Processed 50% of data", {
    progress: 0.5,
    items_processed: 50,
    items_total: 100
  })
  
  # More work
  sleep(1)
  
  # Example GraphQL query
  result_thread = graphql('
    {
      searchUser(_limit: 5, _where: {active: {_eq: true}}) {
        euuid
        name
        type
      }
    }
  ')
  
  result = result_thread.value
  users = result.dig("data", "searchUser") || []
  info("Found active users", { count: users.length })
  
  # Final report
  report("Task completed successfully", {
    items_processed: 100,
    duration_seconds: 2,
    status: "complete"
  })
rescue => e
  error("Task processing failed", { error: e.message })
  raise
end

def main
  # Initialize EYWA client
  open_pipe
  sleep(0.1)
  
  info("Robot started")
  
  begin
    process_task
    info("All operations completed")
    close_task(SUCCESS)
  rescue => e
    error("Task failed", {
      error: e.message,
      type: e.class.name
    })
    close_task(ERROR)
  end
end

# Run the robot
main

File: bb/examples/simple_robot.clj
#!/usr/bin/env bb

(require '[babashka.deps :as deps])
(deps/add-classpath "src")

(require '[eywa.client :as eywa]
         '[clojure.core.async :as async])

(defn -main []
  ;; Start the EYWA client
  (eywa/open-pipe) ; or (eywa/start)

  (try
    ;; Simple info logging
    (eywa/info "Robot started")

    ;; Get current task (if running in task context)
    (async/go
      (let [task-result (async/<! (eywa/get-task))]
        (if-let [task (:result task-result)]
          (do
            (eywa/info "Processing task" {:task-id (:euuid task)})
            ;; Update status
            (eywa/update-task eywa/PROCESSING)

            ;; Do some work...
            (Thread/sleep 1000)

            ;; Report progress
            (eywa/report "Processed 50%" :data {:progress 0.5})

            ;; Do more work...
            (Thread/sleep 1000)

            ;; Success!
            (eywa/report "Task completed" :data {:items-processed 100})
            (eywa/close-task eywa/SUCCESS))
          (do
            ;; Not in task context, just demonstrate features
            (eywa/warn "No task context, running demo mode")

            ;; Show GraphQL usage
            (let [result (async/<! (eywa/graphql "{ searchUser(_limit: 2) { name type } }"))]
              (if (instance? Exception result)
                (eywa/error "GraphQL failed" {:error (.getMessage result)})
                (eywa/info "Found users" {:users (get-in result ["data" "searchUser"])})))

            (eywa/close-task eywa/SUCCESS)))))

    ;; Keep the process alive for async operations
    (Thread/sleep 3000)

    (catch Exception e
      (eywa/exception "Unhandled error" {:error (.getMessage e)
                                         :type (type e)})
      (eywa/close-task eywa/ERROR))))

(-main)
